src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -326,20 +336,55 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
     }
 
     def annotations(): List[Tree] = {
-      //var annots = new ListBuffer[Tree]
+      var annots = new ListBuffer[Tree]
       while (in.token == AT) {
         in.nextToken()
-        annotation()
+        annots += annotation()
       }
-      List() // don't pass on annotations for now
+      annots.toList
     }
 
     /** Annotation ::= TypeName [`(` AnnotationArgument {`,` AnnotationArgument} `)`]
      */
-    def annotation() {
-      qualId()
-      if (in.token == LPAREN) { skipAhead(); accept(RPAREN) }
-      else if (in.token == LBRACE) { skipAhead(); accept(RBRACE) }
+    def annotation(): Tree = {
+      def annArg(): Tree = {
+        def annVal(): Tree = {
```

Comment:
I also think an eager warning is too strict, there is probably a lot of Java code in mixed projects that (unnecessarily / unknowingly) goes through the Scala compiler.

It's always so easy to find a good reason for yet another compiler flag.. Maybe we can avoid that somehow, I don't have an idea right now though. I was thinking of a meta-annotation (or language import) that people could put on their Java code if they want strict Java annotation parsing. But because we're early in the compiler, there's no symbolic information available yet, so we cannot reliably identify a meta-annotation or import.

How to represent annotations with arguments that we cannot constant fold, I'm not sure. There are two pieces to the puzzle: one is arguments we cannot parse (because we don't parse Java expressions), and the other is expressions that we cannot constant fold. Maybe we can exclude the second because we only parse expressions that we can also constant fold. Just for reference, the following works (with the current form of this PR):
```
public @interface A {
  int value();
  public static class T {
    public static final int x = 1;
    @A(x) void t2() { }
  }
}
```

We could drop annotations entirely, or use `ErroneousAnnotation` (which already exists) without issuing an error. I think my preference is to drop them and have a strict mode to get warnings.